rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,127 @@
+
+Furthermore, the shared `EncodingConfig` as it is implemented today is concerned with three tasks: Reshaping an event (including/excluding fields), serializing an event to a byte message, and framing/batching events. In accordance with the functionality that the decoding side provides, we want to keep these concepts separate for a simpler mental model. In fact, we want to separate reshaping from codecs entirely and move it to the responsibilities of the schema work as outlined in the next section.
+
+## Cross cutting concerns
+
+The codec work is very interweaved with the ongoing [schema work](https://github.com/vectordotdev/vector/pull/9388). It introduces the concept of "mappers" which understand how to transform an event at runtime based on the schema from its source and destination within Vector's topology.
+
+We want to hand over the reshaping responsibilities to the schema, and provide a transition layer to support existing reshaping capabilities of sinks until schema mappers are implemented.
+
+Another related change is `StandardEncodings` which was introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent implementation of common codecs, and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.
+
+## Scope
+
+### In scope
+
+- Common support structures for encoding/framing
+- Set of commonly used encoding (text, JSON, ...) and framing (newline-delimited, octet-counting, ...) methods
+- Strategy for integrating encoding/framing with sinks
+
+### Out of scope
+
+- Integration with schema system (automatic transformation of event fields depending)
+
+## Pain
+
+Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.
+
+Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.
+
+## Proposal
+
+### User Experience
+
+Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.
```

Comment:
Should they be able to set these on every type of sink? I'm curious if it would ever make sense to set `framing` on something like the `datadog_logs` sink.

Reply:
Clarified in 6c19b6bf7b80b9adc1971cae75db08d5f4f90b83.